### ISSUE TITLE
Alternate implemention of FUSE for B2

### ISSUE THREAD

**Issue by sondree** [Cfretz244/b2fs#1]
Hi!

I made a B2 FUSE implementation a few months back https://github.com/sondree/b2_fuse 

Have you seen it? Is this what you are trying to achieve? 

Cheers
Sondre

**Comment by Cfretz244** [comment]
Sondre,
This is exactly what I was hoping to accomplish! I was actually about to start work on my end today, so your timing is perfect. The only problem is that I was planning on using this filesystem to host content for a media server, so, as your implementation states that it caches files in memory, that may not work for my particular use case as my files will be fairly large (2-4 gb). Is your code configurable to only cache a certain amount of data in advance, or does it always attempt the cache the whole file?

**Comment by Cfretz244** [comment]
I'm planning on primarily using this for a digital ocean vps with only like 2 gb of ram. Since my use case will be for a media server (Plex in particular), it requires read-only, typically sequential (unless the user intentionally jumps around in the stream), access, and I had been planning on having my code selectively buffer files on a configurable rolling basis. For instance, always staying like 100 mb ahead of the stream and immediately freeing data after it has been served.

**Comment by Cfretz244** [comment]
Out of curiosity, what kind of latency and throughput figures have you been experiencing? That's one thing I've been a little nervous about, as I've never used b2 before.

**Comment by sondree** [comment]
Yes, my implementation caches files in memory. This is caused by the requirement to upload files in a single upload, essentially requiring the entire file to be present at upload time. Since then the B2 API has changed and now allows partial uploads, which could be used to solve the issue you present. 

This seems like an attractive use-case (and one it would be interesting to pursue). One of my current issue is one of application design, I am at a standstill on how to proceed forward. My use-case is a replacement for general cloud storage apps, such as Dropbox. For this  I require a local copy synchronized using the cloud and typically this is mostly small files (< 100mb),

My testing so far indicates that the performance reading many (small) files is bad, mainly due to latency issues. I have attempted to rectify this by multi threading uploads, which gives good performance. This is on a separate branch of the repo. This only partly solves my issue (of many small files), while introducing potential race conditions in the code. Other operations suffer similar performance issues. Typically there is fixed time for each upload of around 1.5-2.5 second, this quickly adds up with many files. Throughput is better, I have been achieving 1.7MB/s or so on uploads, this would be the limit for your use-case.

I would be most interested in collaborating on this project. Looking forward to hearing back from you.

**Comment by Cfretz244** [comment]
Sondre,
I've looked over your code, and it all seems fairly logical. And yeah, unfortunately, it won't work for my use case as it downloads and caches the entire file on reads whereas I would require streaming. As for uploads, you're right, the B2 requirement of knowing the full file size in advance makes it seem impossible to do without caching. The API refers to an upcoming partial upload system, but it says it isn't available yet. The only alternative to an in memory cache I can think of with the current API is to write the data out to disk locally in a tempfile, and then stream the data to B2 upon a release, flush or fsync. Of course the full file will always have to be re-written each time as well. It's a shame the B2 api is limited in this way, although I guess it's understandable.

**Comment by Cfretz244** [comment]
To be honest, I was thinking that this project would be an excellent way for me to brush up on my C skills, so I think I'll probably fool around with that for a while on my own, but knowing how C is, I'll probably get frustrated with it pretty quickly, in which case I'd definitely be interested in doing a collaborative thing.

**Comment by sondree** [comment]
You should have a look at the other branch of the project. In this branch I am implementing a filesystem backed driver. This is similar to what Dropbox does. Currently I am considering if it makes sense to read directly from disk instead of first loading into memory when uploading. This is quite close to what you wish to do, when Backblaze gets around to releasing their multi-part upload feature. 

The major change in the new branch is the more or less instantaneous access that can be had when you keep a local disk copy. There are still some quirks though, as many applications flush and release the files repeatedly leading to multiple successive uploads of the same file.

C should not be a requirement for the performance you require. The bare minimum example overlay of a local filesystem can get 60MB/s + using python wrapped FUSE calls. This is the case for the "local copy" version of the FUSE driver.

**Comment by CafeLungo** [comment]
I'd just like to add my vote for my interest in both of your guys projects. I, too, am looking to potentially host media files on B2, and stream them for local consumption if the performance is good enough. I've played with Sondree's impl, and it is indeed quite slow, but some of that might be due to limitations of the B2 API. For example, the file list calls cannot filter on folders, so it seems very inefficient for listing files, etc.

**Comment by sondree** [comment]
The most important feature required for streaming would be a good implementation using the large file API. I think we would need some sort of lookahead as well, in order to stay ahead of the application reading the data. What kind of data structure to use when storing the partial file would  be non-trivial, you would essentially have to have something that would handle sparse files or files where only segments of the data are present. 

Bulk download speed is quite decent on B2 so assuming one can get around the latency issues it might be possible to store large media files there.

**Comment by Cfretz244** [comment]
Yeah, my plan for file reading is to split every file up into configurable sized chunks and have a thread pool handling a job queue of chunk requests. The requested data chunks would be stored in a variant on a binary tree, allowing O(log n) insertions, deletions, and searches, and efficient iteration (O(1) access to the next entry, given a particular entry). Basically it's an AVL tree that inserts based off of keys (chunk numbers in this case), and also maintains an in order linked list of its nodes.  When an open request is made for a file, the download pool would race ahead, making chunk requests up to a configurable distance ahead of the reader, and read operations would simply iterate through the buffered chunks. For residency calculation, I'm planning on using a basic growing bitmap. For eviction, I was planning on maintaining "current location" markers for each reader, evicting chunks based on a longest forward distance policy. I'd also like for the user to be able to specify maximum memory usage, as I'll be using this on a DigitalOcean vps which won't have a whole lot of ram.

**Comment by Cfretz244** [comment]
Yeah, unfortunately the B2 api is really quite limited at the moment. Even a straightforward mv operation currently requires downloading the file, deleting the original, and reuploading. I haven't decided if I'm going to support it yet. Another example is for uploading files. You're required to specify the content length for any uploads, which in this case basically means the file needs to be completed before it can be uploaded. I'm currently planning on spilling writes to disk, and then starting the upload once the user closes the file. Haven't decided if I'm going to attempt to allow appending to a file, or simply require that all files are truncated upon opening with write flags.

**Comment by Cfretz244** [comment]
Things will get a little easier when they release the large file API, but last I checked it was still marked as experimental and not to be used.
